flang/lib/Lower/HostAssociations.cpp

```diff
@@ -7,19 +7,430 @@
+                    builder.create<fir::ResultOp>(loc, readLength());
+                  })
+                  .genElse([&]() {
+                    auto undef = builder.create<fir::UndefOp>(loc, idxTy);
+                    builder.create<fir::ResultOp>(loc, undef.getResult());
+                  })
+                  .getResults()[0];
+          nonDeferredLenParams.push_back(len);
+        } else {
+          nonDeferredLenParams.push_back(readLength());
+        }
+      }
+    } else if (isDerivedWithLengthParameters(sym)) {
+      TODO(loc, "host associated derived type allocatable or pointer with "
+                "length parameters");
+    }
+    args.symMap.addSymbol(
+        sym, fir::MutableBoxValue(args.valueInTuple, nonDeferredLenParams, {}));
+  }
+};
+
+/// Class defining how arrays are captured inside internal procedures.
+/// Array are captured via a fir.box<fir.ptr<T>> pointer descriptor that
+/// belongs to the host tuple. This allows capturing lower bounds, which
+/// a non pointer descriptor (fir.box<T>) would not allow.
+class CapturedArrays : public CapturedSymbols<CapturedArrays> {
+
+  // Note: Constant shape arrays are not specialized (their base address would
+  // be sufficient information inside the tuple). They could be specialized in
+  // a later FIR pass, or A CapturedStaticShapeArrays could be added to deal
+  // with them here.
+public:
+  static mlir::Type getType(Fortran::lower::AbstractConverter &converter,
+                            const Fortran::semantics::Symbol &sym) {
+    auto type = converter.genType(sym);
+    assert(type.isa<fir::SequenceType>() && "must be a sequence type");
+    return fir::BoxType::get(fir::PointerType::get(type));
+  }
+
+  static void placeInTuple(const PlaceInTuple &args,
+                           Fortran::lower::AbstractConverter &converter,
+                           const Fortran::semantics::Symbol &sym) {
+    auto &builder = converter.getFirOpBuilder();
+    auto loc = args.loc;
+    fir::MutableBoxValue boxInTuple(args.addrInTuple, {}, {});
+    if (args.hostValue.getBoxOf<fir::BoxValue>() &&
+        Fortran::semantics::IsOptional(sym)) {
+      // The assumed shape optional case need some care because it is illegal to
+      // read the incoming box if it is absent (this would cause segfaults).
+      // Pointer association requires reading the target box, so it can only be
+      // done on present optional. For absent optionals, simply create a
+      // disassociated pointer (it is illegal to inquire about lower bounds or
+      // lengths of optional according to 15.5.2.12 3 (9) and 10.1.11 2 (7)b).
+      auto isPresent = builder.create<fir::IsPresentOp>(
+          loc, builder.getI1Type(), fir::getBase(args.hostValue));
+      builder.genIfThenElse(loc, isPresent)
+          .genThen([&]() {
+            Fortran::lower::associateMutableBoxWithShift(
+                builder, loc, boxInTuple, args.hostValue, /*shift*/ llvm::None);
```

Comment:
This reads a bit funny with the `...BoxWithShift` and a `llvm::None` shift.

Reply:
Fair point, I am not a fan of `associateMutableBoxWithShift` given this really implements the normal pointer assignment with the option of giving new lower bounds to the pointer (optional shift). So I will simply rename it `associateMutableBox`.